# Paghahanda ng Datos

[Orihinal na Notebook mula sa *Data Science: Introduction to Machine Learning for Data Science Python and Machine Learning Studio ni Lee Stott*](https://github.com/leestott/intro-Datascience/blob/master/Course%20Materials/4-Cleaning_and_Manipulating-Reference.ipynb)

## Paggalugad ng impormasyon ng `DataFrame`

> **Layunin ng pag-aaral:** Sa pagtatapos ng bahaging ito, dapat kang maging komportable sa paghahanap ng pangkalahatang impormasyon tungkol sa datos na nakaimbak sa pandas DataFrames.

Kapag na-load mo na ang iyong datos sa pandas, malamang na ito ay nasa anyo ng isang `DataFrame`. Gayunpaman, kung ang data set sa iyong `DataFrame` ay may 60,000 na hanay at 400 na kolum, paano mo sisimulan ang pag-unawa sa kung ano ang iyong pinagtatrabahuhan? Sa kabutihang-palad, ang pandas ay nagbibigay ng ilang maginhawang kasangkapan upang mabilis na makita ang pangkalahatang impormasyon tungkol sa isang `DataFrame` bukod pa sa unang ilang at huling ilang hanay.

Upang galugarin ang kakayahang ito, mag-i-import tayo ng Python scikit-learn library at gagamit ng isang iconic na dataset na pamilyar sa bawat data scientist: ang *Iris* dataset ng British biologist na si Ronald Fisher na ginamit sa kanyang papel noong 1936 na "The use of multiple measurements in taxonomic problems":


In [1]:
import pandas as pd
from sklearn.datasets import load_iris

iris = load_iris()
iris_df = pd.DataFrame(data=iris['data'], columns=iris['feature_names'])

### `DataFrame.shape`
Na-load na natin ang Iris Dataset sa variable na `iris_df`. Bago tayo sumisid sa datos, mahalagang malaman ang bilang ng mga datapoint na mayroon tayo at ang kabuuang laki ng dataset. Kapaki-pakinabang na tingnan ang dami ng datos na ating pinagtatrabahuhan.


In [2]:
iris_df.shape

(150, 4)

Kaya, mayroon tayong 150 na hanay at 4 na kolum ng datos. Ang bawat hanay ay kumakatawan sa isang datapoint at ang bawat kolum ay kumakatawan sa isang tampok na nauugnay sa data frame. Kaya't sa madaling salita, mayroong 150 na datapoint na may tig-4 na tampok bawat isa.

Ang `shape` dito ay isang katangian ng dataframe at hindi isang function, kaya hindi ito nagtatapos sa pares ng mga panaklong.


### `DataFrame.columns`
Ngayon, talakayin natin ang 4 na column ng datos. Ano nga ba ang eksaktong kinakatawan ng bawat isa sa kanila? Ang `columns` na attribute ay magbibigay sa atin ng mga pangalan ng mga column sa dataframe.


In [3]:
iris_df.columns

Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)'],
      dtype='object')

Tulad ng nakikita natin, mayroong apat (4) na kolum. Ang `columns` na katangian ay nagsasabi sa atin ng pangalan ng mga kolum at wala nang iba. Nagiging mahalaga ang katangiang ito kapag nais nating tukuyin ang mga tampok na nilalaman ng isang dataset.


### `DataFrame.info`
Ang dami ng datos (ibinibigay ng `shape` attribute) at ang pangalan ng mga tampok o kolum (ibinibigay ng `columns` attribute) ay nagbibigay ng impormasyon tungkol sa dataset. Ngayon, nais nating mas pag-aralan ang dataset. Ang `DataFrame.info()` na function ay lubos na kapaki-pakinabang para dito.


In [4]:
iris_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
dtypes: float64(4)
memory usage: 4.8 KB


Mula rito, makakagawa tayo ng ilang obserbasyon:  
1. Ang Uri ng Data ng bawat kolum: Sa dataset na ito, lahat ng data ay nakaimbak bilang 64-bit floating-point numbers.  
2. Bilang ng mga Non-Null na halaga: Ang paghawak sa mga null na halaga ay isang mahalagang hakbang sa paghahanda ng data. Ito ay tatalakayin sa mga susunod na bahagi ng notebook.  


### DataFrame.describe()
Halimbawa, mayroon tayong maraming numerikal na datos sa ating dataset. Ang mga unibaryong kalkulasyong estadistikal tulad ng mean, median, quartiles, at iba pa ay maaaring gawin sa bawat kolum nang paisa-isa. Ang `DataFrame.describe()` na function ay nagbibigay sa atin ng estadistikal na buod ng mga numerikal na kolum ng isang dataset.


In [5]:
iris_df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
count         150.000000        150.000000         150.000000        150.000000
mean            5.843333          3.057333           3.758000          1.199333
std             0.828066          0.435866           1.765298          0.762238
min             4.300000          2.000000           1.000000          0.100000
25%             5.100000          2.800000           1.600000          0.300000
50%             5.800000          3.000000           4.350000          1.300000
75%             6.400000          3.300000           5.100000          1.800000
max             7.900000          4.400000           6.900000          2.500000

Ang output sa itaas ay nagpapakita ng kabuuang bilang ng mga data point, mean, standard deviation, minimum, lower quartile (25%), median (50%), upper quartile (75%), at ang maximum na halaga ng bawat column.


### `DataFrame.head`
Sa lahat ng mga nabanggit na function at attribute, mayroon na tayong pangkalahatang ideya tungkol sa dataset. Alam na natin kung ilang data points ang naroon, kung ilang features ang mayroon, ang uri ng data ng bawat feature, at ang bilang ng mga non-null na halaga para sa bawat feature.

Ngayon, oras na para tingnan mismo ang data. Tingnan natin kung ano ang hitsura ng unang ilang row (ang unang ilang data points) ng ating `DataFrame`:


In [6]:
iris_df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

Makikita natin dito sa output ang limang (5) entry ng dataset. Kung titingnan natin ang index sa kaliwa, malalaman natin na ito ang unang limang hanay.


### Ehersisyo:

Mula sa halimbawang ibinigay sa itaas, malinaw na, sa default, ang `DataFrame.head` ay nagbabalik ng unang limang hanay ng isang `DataFrame`. Sa code cell sa ibaba, kaya mo bang alamin kung paano ipakita ang higit sa limang hanay?


In [7]:
# Hint: Consult the documentation by using iris_df.head?

### `DataFrame.tail`
Isa pang paraan ng pagtingin sa datos ay mula sa dulo (sa halip na sa simula). Ang kabaligtaran ng `DataFrame.head` ay ang `DataFrame.tail`, na nagbabalik ng huling limang hanay ng isang `DataFrame`:


In [8]:
iris_df.tail()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

Sa praktika, mahalagang magkaroon ng kakayahang madaling masuri ang unang ilang hanay o ang huling ilang hanay ng isang `DataFrame`, lalo na kung naghahanap ka ng mga outlier sa mga nakaayos na dataset.

Ang lahat ng mga function at attribute na ipinakita sa itaas gamit ang mga halimbawa ng code ay tumutulong sa atin na magkaroon ng ideya at pakiramdam tungkol sa datos.

> **Mahalagang Punto:** Kahit sa simpleng pagtingin lamang sa metadata tungkol sa impormasyon sa isang DataFrame o sa unang at huling ilang halaga nito, maaari ka nang magkaroon ng agarang ideya tungkol sa laki, hugis, at nilalaman ng datos na iyong hinaharap.


### Nawawalang Datos
Tuklasin natin ang tungkol sa nawawalang datos. Ang nawawalang datos ay nangyayari kapag walang halaga na nakaimbak sa ilang mga kolum.

Halimbawa: sabihin nating may isang tao na masyadong sensitibo tungkol sa kanyang timbang at hindi pinupunan ang field para sa timbang sa isang survey. Sa ganitong kaso, ang halaga ng timbang para sa taong iyon ay magiging nawawala.

Kadalasan, sa mga totoong dataset sa mundo, madalas na may mga nawawalang halaga.

**Paano Hinahawakan ng Pandas ang Nawawalang Datos**

Hinahawakan ng Pandas ang nawawalang mga halaga sa dalawang paraan. Ang una ay nakita mo na sa mga naunang seksyon: `NaN`, o Not a Number. Ito ay isang espesyal na halaga na bahagi ng IEEE floating-point specification at ginagamit lamang upang ipahiwatig ang nawawalang floating-point na mga halaga.

Para sa mga nawawalang halaga na hindi floats, ginagamit ng pandas ang Python na `None` object. Bagama't maaaring nakakalito na makatagpo ng dalawang magkaibang uri ng halaga na nagsasabi ng halos parehong bagay, may mga makatwirang programmatic na dahilan para sa ganitong disenyo. Sa praktika, ang ganitong paraan ay nagbibigay-daan sa pandas na maghatid ng balanseng solusyon para sa karamihan ng mga kaso. Gayunpaman, parehong `None` at `NaN` ay may mga limitasyon na kailangan mong tandaan kaugnay sa kung paano sila maaaring gamitin.


### `None`: non-float na nawawalang data
Dahil ang `None` ay nagmula sa Python, hindi ito maaaring gamitin sa mga NumPy at pandas arrays na hindi may data type na `'object'`. Tandaan, ang mga NumPy arrays (at ang mga data structure sa pandas) ay maaaring maglaman lamang ng isang uri ng data. Ito ang nagbibigay sa kanila ng napakalaking kapangyarihan para sa malakihang data at computational na gawain, ngunit nililimitahan din nito ang kanilang kakayahang mag-adjust. Ang ganitong mga arrays ay kailangang mag-upcast sa “pinakamababang karaniwang denominador,” ang data type na kayang maglaman ng lahat sa array. Kapag ang `None` ay nasa array, nangangahulugan ito na ikaw ay nagtatrabaho gamit ang mga Python objects.

Upang makita ito sa aktwal, isaalang-alang ang sumusunod na halimbawa ng array (pansinin ang `dtype` nito):


In [9]:
import numpy as np

example1 = np.array([2, None, 6, 8])
example1

array([2, None, 6, 8], dtype=object)

Ang realidad ng paggamit ng upcast na mga data type ay may dalawang epekto. Una, ang mga operasyon ay isasagawa sa antas ng interpreted na Python code sa halip na compiled na NumPy code. Sa madaling salita, nangangahulugan ito na ang anumang operasyon na may kinalaman sa `Series` o `DataFrames` na may `None` ay magiging mas mabagal. Bagamat maaaring hindi mo mapansin ang epekto nito sa performance, para sa malalaking dataset, maaari itong maging isang isyu.

Ang pangalawang epekto ay nagmumula sa una. Dahil ang `None` ay karaniwang nagdadala ng `Series` o `DataFrame` pabalik sa mundo ng vanilla Python, ang paggamit ng mga NumPy/pandas aggregation tulad ng `sum()` o `min()` sa mga array na naglalaman ng `None` na halaga ay karaniwang magdudulot ng error:


In [10]:
example1.sum()

TypeError: ignored

### `NaN`: nawawalang float na halaga

Sa kaibahan sa `None`, sinusuportahan ng NumPy (at samakatuwid ng pandas) ang `NaN` para sa mabilis, vectorized na mga operasyon at ufuncs. Ang masamang balita ay anumang arithmetic na ginawa sa `NaN` ay palaging nagreresulta sa `NaN`. Halimbawa:


In [11]:
np.nan + 1

nan

In [12]:
np.nan * 0

nan

Ang magandang balita: ang mga pagsasama-sama na tumatakbo sa mga array na may `NaN` sa loob nito ay hindi nagdudulot ng mga error. Ang masamang balita: ang mga resulta ay hindi palaging kapaki-pakinabang:


In [13]:
example2 = np.array([2, np.nan, 6, 8]) 
example2.sum(), example2.min(), example2.max()

(nan, nan, nan)

### Ehersisyo:


In [11]:
# What happens if you add np.nan and None together?


### `NaN` at `None`: mga null na halaga sa pandas

Kahit na ang `NaN` at `None` ay maaaring kumilos nang medyo magkaiba, ang pandas ay ginawa upang hawakan ang mga ito nang magkapareho. Upang makita kung ano ang ibig naming sabihin, isaalang-alang ang isang `Series` ng mga integer:


In [15]:
int_series = pd.Series([1, 2, 3], dtype=int)
int_series

0    1
1    2
2    3
dtype: int64

### Ehersisyo:


In [16]:
# Now set an element of int_series equal to None.
# How does that element show up in the Series?
# What is the dtype of the Series?


Sa proseso ng pag-upcast ng mga uri ng data upang maitatag ang pagkakapareho ng data sa `Series` at `DataFrame`s, kusang pinapalitan ng pandas ang mga nawawalang halaga sa pagitan ng `None` at `NaN`. Dahil sa katangiang ito ng disenyo, makatutulong na isipin ang `None` at `NaN` bilang dalawang magkaibang uri ng "null" sa pandas. Sa katunayan, ang ilang pangunahing mga pamamaraan na gagamitin mo upang harapin ang mga nawawalang halaga sa pandas ay sumasalamin sa ideyang ito sa kanilang mga pangalan:

- `isnull()`: Gumagawa ng Boolean mask na nagpapakita ng mga nawawalang halaga
- `notnull()`: Kabaligtaran ng `isnull()`
- `dropna()`: Nagbabalik ng na-filter na bersyon ng data
- `fillna()`: Nagbabalik ng kopya ng data na may napunan o na-impute na mga nawawalang halaga

Mahalagang matutunan at maging pamilyar sa mga pamamaraang ito, kaya talakayin natin ang bawat isa nang mas malalim.


### Pagtukoy sa mga null na halaga

Ngayon na nauunawaan na natin ang kahalagahan ng mga nawawalang halaga, kailangan muna nating tukuyin ang mga ito sa ating dataset bago ito harapin.  
Ang parehong `isnull()` at `notnull()` ang iyong pangunahing mga pamamaraan para matukoy ang null na data. Ang dalawa ay nagbabalik ng Boolean masks sa iyong data.


In [17]:
example3 = pd.Series([0, np.nan, '', None])

In [18]:
example3.isnull()

0    False
1     True
2    False
3     True
dtype: bool

Tingnan mong mabuti ang output. Mayroon bang anumang bagay na nakakagulat sa iyo? Bagama't ang `0` ay isang arithmetic null, ito ay isang ganap na wastong integer at tinatrato ito ng pandas bilang ganoon. Ang `''` naman ay medyo mas maselan. Bagama't ginamit natin ito sa Seksyon 1 upang kumatawan sa isang walang laman na string na halaga, ito ay isang string object pa rin at hindi itinuturing na null ayon sa pandas.

Ngayon, baliktarin natin ito at gamitin ang mga pamamaraang ito sa paraang mas malapit sa aktwal na paggamit mo sa kanila. Maaari mong gamitin ang Boolean masks nang direkta bilang isang ``Series`` o ``DataFrame`` index, na maaaring maging kapaki-pakinabang kapag sinusubukang magtrabaho sa mga nakahiwalay na nawawalang (o naroroon) na mga halaga.

Kung nais nating makuha ang kabuuang bilang ng mga nawawalang halaga, maaari lamang tayong mag-sum sa mask na ginawa ng `isnull()` na pamamaraan.


In [19]:
example3.isnull().sum()

2

### Ehersisyo:


In [20]:
# Try running example3[example3.notnull()].
# Before you do so, what do you expect to see?


**Pangunahing aral**: Parehong ang mga pamamaraan na `isnull()` at `notnull()` ay nagbibigay ng magkatulad na resulta kapag ginamit mo ang mga ito sa mga DataFrame: ipinapakita nila ang mga resulta at ang index ng mga resulta, na makakatulong sa iyo nang malaki habang pinoproseso mo ang iyong data.


### Pagtugon sa Nawawalang Datos

> **Layunin ng Pag-aaral:** Sa pagtatapos ng bahaging ito, dapat mong malaman kung paano at kailan papalitan o aalisin ang mga null na halaga mula sa mga DataFrame.

Ang mga modelo ng Machine Learning ay hindi kayang direktang magproseso ng nawawalang datos. Kaya, bago ipasa ang datos sa modelo, kailangan nating tugunan ang mga nawawalang halaga.

Ang paraan ng paghawak sa nawawalang datos ay may kasamang mga maseselang tradeoff, at maaaring makaapekto sa iyong huling pagsusuri at mga resulta sa totoong mundo.

Mayroong pangunahing dalawang paraan upang tugunan ang nawawalang datos:

1.   Alisin ang row na naglalaman ng nawawalang halaga  
2.   Palitan ang nawawalang halaga ng ibang halaga  

Tatalakayin natin ang parehong mga pamamaraang ito at ang kanilang mga kalamangan at kahinaan nang detalyado.


### Pag-aalis ng mga null na halaga

Ang dami ng datos na ipinapasa natin sa ating modelo ay may direktang epekto sa performance nito. Ang pag-aalis ng mga null na halaga ay nangangahulugan na binabawasan natin ang bilang ng mga datapoint, at sa gayon ay binabawasan ang laki ng dataset. Kaya, inirerekomenda na alisin ang mga row na may null na halaga kapag ang dataset ay medyo malaki.

Isa pang halimbawa ay kung ang isang partikular na row o column ay may maraming nawawalang halaga. Sa ganitong kaso, maaaring alisin ang mga ito dahil hindi naman sila makakapagdagdag ng malaking halaga sa ating pagsusuri dahil karamihan sa datos para sa row/column na iyon ay nawawala.

Bukod sa pagtukoy ng mga nawawalang halaga, ang pandas ay nagbibigay ng maginhawang paraan upang alisin ang mga null na halaga mula sa `Series` at `DataFrame`s. Upang makita ito sa aksyon, balikan natin ang `example3`. Ang function na `DataFrame.dropna()` ay tumutulong sa pag-aalis ng mga row na may null na halaga.


In [21]:
example3 = example3.dropna()
example3

0    0
2     
dtype: object

Tandaan na dapat itong magmukhang iyong output mula sa `example3[example3.notnull()]`. Ang pagkakaiba dito ay, sa halip na i-index lamang ang mga masked na halaga, inalis ng `dropna` ang mga nawawalang halaga mula sa `Series` na `example3`.

Dahil ang mga DataFrame ay may dalawang dimensyon, nagbibigay ito ng mas maraming opsyon para sa pag-aalis ng data.


In [22]:
example4 = pd.DataFrame([[1,      np.nan, 7], 
                         [2,      5,      8], 
                         [np.nan, 6,      9]])
example4

0    1  2
0  1.0  NaN  7
1  2.0  5.0  8
2  NaN  6.0  9

(Napansin mo ba na ang pandas ay nag-upcast ng dalawang column sa floats upang ma-accommodate ang `NaN`s?)

Hindi mo maaaring tanggalin ang isang solong halaga mula sa isang `DataFrame`, kaya kailangan mong tanggalin ang buong mga row o column. Depende sa ginagawa mo, maaaring mas gusto mong gawin ang isa kaysa sa isa pa, kaya binibigyan ka ng pandas ng mga opsyon para sa pareho. Dahil sa data science, ang mga column ay karaniwang kumakatawan sa mga variable at ang mga row ay kumakatawan sa mga obserbasyon, mas malamang na tanggalin mo ang mga row ng data; ang default na setting para sa `dropna()` ay tanggalin ang lahat ng mga row na may anumang null na halaga:


In [23]:
example4.dropna()

0    1  2
1  2.0  5.0  8

Kung kinakailangan, maaari mong tanggalin ang mga NA na halaga mula sa mga kolum. Gamitin ang `axis=1` upang gawin ito:


In [24]:
example4.dropna(axis='columns')

2
0  7
1  8
2  9

Tandaan na maaaring mawala ang maraming datos na maaaring gusto mong itago, lalo na sa mas maliliit na dataset. Paano kung gusto mo lang tanggalin ang mga row o column na may ilang o kahit lahat ng null na halaga? Maaari mong tukuyin ang mga setting na ito sa `dropna` gamit ang mga parameter na `how` at `thresh`.

Sa default, `how='any'` (kung nais mong suriin ito mismo o makita ang iba pang mga parameter na mayroon ang method, patakbuhin ang `example4.dropna?` sa isang code cell). Maaari mo ring tukuyin ang `how='all'` upang tanggalin lamang ang mga row o column na may lahat ng null na halaga. Palawakin natin ang ating halimbawa ng `DataFrame` upang makita ito sa aksyon sa susunod na ehersisyo.


In [25]:
example4[3] = np.nan
example4

0    1  2   3
0  1.0  NaN  7 NaN
1  2.0  5.0  8 NaN
2  NaN  6.0  9 NaN

> Mahahalagang Punto:  
1. Ang pagtanggal ng mga null na halaga ay magandang ideya lamang kung sapat na kalaki ang dataset.  
2. Maaaring tanggalin ang buong mga row o column kung karamihan sa kanilang data ay nawawala.  
3. Ang `DataFrame.dropna(axis=)` na method ay tumutulong sa pagtanggal ng mga null na halaga. Ang argumentong `axis` ay nagpapahiwatig kung ang mga row o column ang tatanggalin.  
4. Maaaring gamitin ang argumentong `how`. Sa default, ito ay nakatakda sa `any`. Kaya, tinatanggal lamang nito ang mga row/column na mayroong kahit anong null na halaga. Maaari itong itakda sa `all` upang tukuyin na tatanggalin lamang ang mga row/column kung lahat ng halaga ay null.  


### Ehersisyo:


In [22]:
# How might you go about dropping just column 3?
# Hint: remember that you will need to supply both the axis parameter and the how parameter.


Ang parameter na `thresh` ay nagbibigay sa iyo ng mas detalyadong kontrol: itinatakda mo ang bilang ng *hindi-null* na mga halaga na kailangang mayroon ang isang row o column upang mapanatili ito:


In [27]:
example4.dropna(axis='rows', thresh=3)

0    1  2   3
1  2.0  5.0  8 NaN

### Pagpuno ng mga null na halaga

Minsan may kabuluhan na punan ang mga nawawalang halaga gamit ang mga maaaring maging wasto. May ilang mga teknik para punan ang mga null na halaga. Ang una ay ang paggamit ng Kaalaman sa Larangan (kaalaman tungkol sa paksa kung saan nakabatay ang dataset) upang sa paanuman ay matantiya ang mga nawawalang halaga.

Maaari mong gamitin ang `isnull` upang gawin ito nang direkta, ngunit maaaring matrabaho ito, lalo na kung marami kang halagang kailangang punan. Dahil ito ay isang karaniwang gawain sa data science, ang pandas ay nagbibigay ng `fillna`, na nagbabalik ng kopya ng `Series` o `DataFrame` kung saan ang mga nawawalang halaga ay pinalitan ng halagang iyong pinili. Gumawa tayo ng isa pang halimbawa ng `Series` upang makita kung paano ito gumagana sa praktika.


### Datos na Kategorya (Hindi-Numeriko)
Unahin nating talakayin ang hindi-numerikong datos. Sa mga dataset, may mga kolum na naglalaman ng datos na kategorya. Halimbawa: Kasarian, Totoo o Mali, at iba pa.

Sa karamihan ng mga ganitong kaso, pinapalitan natin ang mga nawawalang halaga gamit ang `mode` ng kolum. Halimbawa, mayroon tayong 100 data points kung saan 90 ang nagsabing Totoo, 8 ang nagsabing Mali, at 2 ang walang sagot. Sa ganitong sitwasyon, maaari nating punan ang 2 na nawawala ng Totoo, batay sa kabuuang datos ng kolum.

Muli, maaari rin nating gamitin ang kaalaman sa domain dito. Tingnan natin ang isang halimbawa ng pagpuno gamit ang mode.


In [28]:
fill_with_mode = pd.DataFrame([[1,2,"True"],
                               [3,4,None],
                               [5,6,"False"],
                               [7,8,"True"],
                               [9,10,"True"]])

fill_with_mode

0   1      2
0  1   2   True
1  3   4   None
2  5   6  False
3  7   8   True
4  9  10   True

In [29]:
fill_with_mode[2].value_counts()

True     3
False    1
Name: 2, dtype: int64

In [30]:
fill_with_mode[2].fillna('True',inplace=True)

In [31]:
fill_with_mode

0   1      2
0  1   2   True
1  3   4   True
2  5   6  False
3  7   8   True
4  9  10   True

Tulad ng nakikita natin, napalitan na ang null na halaga. Hindi na kailangang sabihin, maaari tayong magsulat ng kahit ano kapalit ng `'True'` at ito ay mapapalitan.


### Numeric Data
Ngayon, pag-usapan natin ang numeric data. Dito, may dalawang karaniwang paraan ng pagpapalit sa mga nawawalang halaga:

1. Palitan gamit ang Median ng row
2. Palitan gamit ang Mean ng row

Ginagamit natin ang Median kapag ang data ay skewed at may mga outlier. Ito ay dahil ang median ay hindi gaanong naaapektuhan ng mga outlier.

Kapag ang data ay na-normalize, maaari nating gamitin ang mean, dahil sa ganitong kaso, ang mean at median ay halos magkapareho.

Una, kumuha tayo ng isang column na may normal na distribusyon at punan natin ang nawawalang halaga gamit ang mean ng column.


In [32]:
fill_with_mean = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [np.nan,4,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_mean

0  1  2
0 -2.0  0  1
1 -1.0  2  3
2  NaN  4  5
3  1.0  6  7
4  2.0  8  9

Ang mean ng kolum ay


In [33]:
np.mean(fill_with_mean[0])

0.0

In [34]:
fill_with_mean[0].fillna(np.mean(fill_with_mean[0]),inplace=True)
fill_with_mean

0  1  2
0 -2.0  0  1
1 -1.0  2  3
2  0.0  4  5
3  1.0  6  7
4  2.0  8  9

Tulad ng nakikita natin, ang nawawalang halaga ay napalitan ng kanyang mean.


Ngayon subukan natin ang isa pang dataframe, at sa pagkakataong ito ay papalitan natin ang mga None na halaga ng median ng kolum.


In [35]:
fill_with_median = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [0,np.nan,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_median

0    1  2
0 -2  0.0  1
1 -1  2.0  3
2  0  NaN  5
3  1  6.0  7
4  2  8.0  9

Ang median ng ikalawang kolum ay


In [36]:
fill_with_median[1].median()

4.0

In [37]:
fill_with_median[1].fillna(fill_with_median[1].median(),inplace=True)
fill_with_median

0    1  2
0 -2  0.0  1
1 -1  2.0  3
2  0  4.0  5
3  1  6.0  7
4  2  8.0  9

Tulad ng nakikita natin, ang NaN na halaga ay napalitan ng median ng kolum


In [38]:
example5 = pd.Series([1, np.nan, 2, None, 3], index=list('abcde'))
example5

a    1.0
b    NaN
c    2.0
d    NaN
e    3.0
dtype: float64

Maaari mong punan ang lahat ng mga null na entry gamit ang isang halaga, tulad ng `0`:


In [39]:
example5.fillna(0)

a    1.0
b    0.0
c    2.0
d    0.0
e    3.0
dtype: float64

> Mahahalagang Punto:
1. Ang paglalagay ng nawawalang halaga ay dapat gawin kapag kaunti lamang ang datos o mayroong estratehiya para punan ang nawawalang datos.
2. Maaaring gamitin ang kaalaman sa larangan upang punan ang nawawalang halaga sa pamamagitan ng pagtantya sa mga ito.
3. Para sa Categorical na datos, kadalasan, ang nawawalang halaga ay pinapalitan ng mode ng kolum.
4. Para sa numeric na datos, ang nawawalang halaga ay karaniwang pinupunan gamit ang mean (para sa mga normalisadong dataset) o ang median ng mga kolum.


### Ehersisyo:


In [40]:
# What happens if you try to fill null values with a string, like ''?


In [41]:
example5.fillna(method='ffill')

a    1.0
b    1.0
c    2.0
d    2.0
e    3.0
dtype: float64

Maaari mo ring **back-fill** upang ipalaganap ang susunod na wastong halaga pabalik upang punan ang null:


In [42]:
example5.fillna(method='bfill')

a    1.0
b    2.0
c    2.0
d    3.0
e    3.0
dtype: float64

Tulad ng maaari mong hulaan, gumagana ito nang pareho sa mga DataFrame, ngunit maaari ka ring magtakda ng isang `axis` kung saan pupunan ang mga null na halaga:


In [43]:
example4

0    1  2   3
0  1.0  NaN  7 NaN
1  2.0  5.0  8 NaN
2  NaN  6.0  9 NaN

In [44]:
example4.fillna(method='ffill', axis=1)

0    1    2    3
0  1.0  1.0  7.0  7.0
1  2.0  5.0  8.0  8.0
2  NaN  6.0  9.0  9.0

Pansinin na kapag walang nakaraang halaga na magagamit para sa forward-filling, nananatili ang null na halaga.


### Ehersisyo:


In [45]:
# What output does example4.fillna(method='bfill', axis=1) produce?
# What about example4.fillna(method='ffill') or example4.fillna(method='bfill')?
# Can you think of a longer code snippet to write that can fill all of the null values in example4?


Maaari kang maging malikhain sa paggamit ng `fillna`. Halimbawa, balikan natin ang `example4`, ngunit sa pagkakataong ito, punan natin ang mga nawawalang halaga gamit ang average ng lahat ng mga halaga sa `DataFrame`:


In [46]:
example4.fillna(example4.mean())

0    1  2   3
0  1.0  5.5  7 NaN
1  2.0  5.0  8 NaN
2  1.5  6.0  9 NaN

Pansinin na ang column 3 ay wala pa ring halaga: ang default na direksyon ay punan ang mga halaga nang pa-row.

> **Punto:** Maraming paraan upang harapin ang mga nawawalang halaga sa iyong mga dataset. Ang partikular na estratehiya na gagamitin mo (pag-aalis ng mga ito, pagpapalit ng mga ito, o kahit paano mo papalitan ang mga ito) ay dapat nakabatay sa mga detalye ng datos na iyon. Mas magkakaroon ka ng mas mahusay na pag-unawa kung paano harapin ang mga nawawalang halaga habang mas madalas kang humawak at makipag-ugnayan sa mga dataset.


### Pag-encode ng Kategoryal na Data

Ang mga modelo ng machine learning ay gumagana lamang sa mga numero at anumang uri ng numerikong data. Hindi nito kayang tukuyin ang pagkakaiba ng Oo at Hindi, ngunit kaya nitong tukuyin ang pagkakaiba ng 0 at 1. Kaya, pagkatapos punan ang mga nawawalang halaga, kailangan nating i-encode ang kategoryal na data sa isang numerikong anyo upang maunawaan ito ng modelo.

Ang pag-encode ay maaaring gawin sa dalawang paraan. Tatalakayin natin ang mga ito sa susunod.


**PAG-ENCODE NG LABEL**

Ang pag-encode ng label ay ang proseso ng pag-convert ng bawat kategorya sa isang numero. Halimbawa, sabihin nating mayroon tayong dataset ng mga pasahero ng eroplano at may isang column na naglalaman ng kanilang klase mula sa mga sumusunod ['business class', 'economy class', 'first class']. Kapag in-encode ang mga label nito, ito ay magiging [0,1,2]. Tingnan natin ang isang halimbawa gamit ang code. Dahil pag-aaralan natin ang `scikit-learn` sa mga susunod na notebook, hindi natin ito gagamitin dito.


In [47]:
label = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
label

ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class

Upang maisagawa ang label encoding sa unang kolum, kailangan muna nating ilarawan ang isang mapping mula sa bawat klase patungo sa isang numero, bago palitan


In [48]:
class_labels = {'business class':0,'economy class':1,'first class':2}
label['class'] = label['class'].replace(class_labels)
label

ID  class
0  10      0
1  20      2
2  30      1
3  40      1
4  50      1
5  60      0

Tulad ng nakikita natin, ang resulta ay tumutugma sa inaasahan natin. Kaya, kailan natin ginagamit ang label encoding? Ginagamit ang label encoding sa isa o pareho sa mga sumusunod na sitwasyon:  
1. Kapag ang bilang ng mga kategorya ay malaki  
2. Kapag ang mga kategorya ay may pagkakasunod-sunod.  


**ONE HOT ENCODING**

Isa pang uri ng encoding ay ang One Hot Encoding. Sa ganitong uri ng encoding, ang bawat kategorya ng isang column ay nagiging hiwalay na column, at ang bawat datapoint ay magkakaroon ng 0 o 1 depende kung kabilang ito sa kategoryang iyon. Kaya, kung mayroong n na magkakaibang kategorya, n na column ang idadagdag sa dataframe.

Halimbawa, kunin natin ang parehong halimbawa ng klase ng eroplano. Ang mga kategorya ay: ['business class', 'economy class', 'first class']. Kaya, kung gagamit tayo ng one hot encoding, ang sumusunod na tatlong column ay idadagdag sa dataset: ['class_business class', 'class_economy class', 'class_first class'].


In [49]:
one_hot = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
one_hot

ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class

Gawin natin ang one hot encoding sa unang kolum


In [50]:
one_hot_data = pd.get_dummies(one_hot,columns=['class'])

In [51]:
one_hot_data

ID  class_business class  class_economy class  class_first class
0  10                     1                    0                  0
1  20                     0                    0                  1
2  30                     0                    1                  0
3  40                     0                    1                  0
4  50                     0                    1                  0
5  60                     1                    0                  0

Ang bawat one-hot encoded na kolum ay naglalaman ng 0 o 1, na nagsasaad kung ang kategoryang iyon ay umiiral para sa datapoint na iyon.


Kailan ginagamit ang one hot encoding? Ginagamit ang one hot encoding sa isa o pareho sa mga sumusunod na sitwasyon:

1. Kapag ang bilang ng mga kategorya at ang laki ng dataset ay mas maliit.
2. Kapag ang mga kategorya ay walang partikular na pagkakasunod-sunod.


> Mahahalagang Punto:
1. Ang encoding ay ginagawa upang ma-convert ang hindi numerikong datos sa numerikong datos.
2. May dalawang uri ng encoding: Label encoding at One Hot encoding, na maaaring isagawa batay sa pangangailangan ng dataset.


## Pag-aalis ng dobleng datos

> **Layunin ng pag-aaral:** Sa pagtatapos ng bahaging ito, dapat kang maging komportable sa pag-identipika at pag-aalis ng mga dobleng halaga mula sa mga DataFrame.

Bukod sa nawawalang datos, madalas kang makakakita ng dobleng datos sa mga aktwal na dataset. Sa kabutihang-palad, nagbibigay ang pandas ng madaling paraan para matukoy at maalis ang mga dobleng entry.


### Pagtukoy ng mga duplicate: `duplicated`

Madali mong matutukoy ang mga duplicate na halaga gamit ang `duplicated` na method sa pandas, na nagbabalik ng Boolean mask na nagpapakita kung ang isang entry sa isang `DataFrame` ay duplicate ng naunang entry. Gumawa tayo ng isa pang halimbawa ng `DataFrame` upang makita ito sa aksyon.


In [52]:
example6 = pd.DataFrame({'letters': ['A','B'] * 2 + ['B'],
                         'numbers': [1, 2, 1, 3, 3]})
example6

letters  numbers
0       A        1
1       B        2
2       A        1
3       B        3
4       B        3

In [53]:
example6.duplicated()

0    False
1    False
2     True
3    False
4     True
dtype: bool

### Pagtanggal ng mga duplicate: `drop_duplicates`
Ang `drop_duplicates` ay simpleng nagbabalik ng kopya ng data kung saan lahat ng mga halagang `duplicated` ay `False`:


In [54]:
example6.drop_duplicates()

letters  numbers
0       A        1
1       B        2
3       B        3

Parehong `duplicated` at `drop_duplicates` ay default na isaalang-alang ang lahat ng mga column ngunit maaari mong tukuyin na suriin lamang nila ang isang subset ng mga column sa iyong `DataFrame`:


In [55]:
example6.drop_duplicates(['letters'])

letters  numbers
0       A        1
1       B        2


---

**Paunawa**:  
Ang dokumentong ito ay isinalin gamit ang AI translation service na [Co-op Translator](https://github.com/Azure/co-op-translator). Bagama't sinisikap naming maging tumpak, pakitandaan na ang mga awtomatikong pagsasalin ay maaaring maglaman ng mga pagkakamali o hindi pagkakatugma. Ang orihinal na dokumento sa orihinal nitong wika ang dapat ituring na opisyal na sanggunian. Para sa mahalagang impormasyon, inirerekomenda ang propesyonal na pagsasalin ng tao. Hindi kami mananagot sa anumang hindi pagkakaunawaan o maling interpretasyon na maaaring magmula sa paggamit ng pagsasaling ito.
